src/remote/remote.go

```diff
@@ -265,10 +261,45 @@ func (c *Client) Build(tid int, target *core.BuildTarget) (*core.BuildMetadata,
 	return metadata, nil
 }
 
+// build implements the actual build of a target.
+func (c *Client) build(tid int, target *core.BuildTarget) (*core.BuildMetadata, *pb.ActionResult, *pb.Digest, error) {
+	needStdout := target.PostBuildFunction != nil
+	// If we're gonna stamp the target, first check the unstamped equivalent that we store results under.
+	// This implements the rules of stamp whereby we don't force rebuilds every time e.g. the SCM revision changes.
+	if target.Stamp {
+		if command, digest, err := c.buildAction(target, false, false); err == nil {
+			if metadata, ar := c.maybeRetrieveResults(tid, target, command, digest, needStdout); metadata != nil {
+				return metadata, ar, digest, nil
+			}
+		}
+	}
+	command, digest, err := c.buildAction(target, false, target.Stamp)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	metadata, ar, err := c.execute(tid, target, command, digest, target.BuildTimeout, false, needStdout)
+	if err != nil {
+		return metadata, ar, digest, err
+	}
+	if target.Stamp {
+		// Store results under unstamped digest too.
+		_, unstampedDigest, _ := c.buildAction(target, false, false)
```

Comment:
This might fail in some really funky ways

Reply:
Have redone a bit so as not to ignore errors. Also not recalculating the digests this time for mad perf gainz